snapshot/src/snap.rs

```diff
@@ -227,6 +241,63 @@ impl Snap {
     }
 }
 
+pub struct SnapReader {
+    sizes: Vec<i32>,
+}
+
+fn read_int_err<W: Warn<Warning>>(p: &mut Unpacker, w: &mut W, e: Error) -> Result<i32, Error> {
+    p.read_int(wrap(w)).map_err(|_| e)
+}
+
+impl SnapReader {
+    pub fn new() -> SnapReader {
+        SnapReader { sizes: Vec::new() }
+    }
+    pub fn read<W>(&mut self, warn: &mut W, snap: Snap, p: &mut Unpacker) -> Result<Snap, Error>
+        where W: Warn<Warning>
+    {
+        self.sizes.clear();
+        let header = SnapHeader::decode(warn, p)?;
+        let mut prev_offset = None;
+        for _ in 0..header.num_items {
+            let offset = p.read_int(wrap(warn)).unwrap();
+            if let Some(prev) = prev_offset {
+                if prev > offset {
+                    return Err(Error::InvalidOffset);
+                }
+                self.sizes.push(offset - prev);
+            } else if offset != 0 {
+                // First offset must be 0
+                return Err(Error::InvalidOffset);
+            }
+            prev_offset = Some(offset);
+        }
+        if let Some(last) = prev_offset {
+            if last > header.data_size {
+                return Err(Error::InvalidOffset);
+            }
+            self.sizes.push(header.data_size - last)
+        }
+
+        let mut builder = snap.recycle();
+        for size in &self.sizes {
```

Comment:
```suggestion
        for &size in &self.sizes {
```
gets rid of the need to dereference below.